### DISCUSSION TITLE
Opengrok Index has failed to create data/historycache for a Git repository

### DISCUSSION THREAD

**Discussion by prabhu1084** [oracle/opengrok#4393] [General]
Hi All,

I have executed the below indexing command (with the -H option) but it is not creating the historycache.

**Indexing command:**
> java \
> "${JAVA_OPTS:--Xmx16g}" \ 
> -server \ 
> -XX:-UseGCOverheadLimit  \ 
> -jar /mnt1/share/opengrok/dist/lib/opengrok.jar \ 
> -v \ 
> -c /usr/local/bin/ctags \ 
> -s /mnt1/share/opengrok/instances/PROJECT/src \ 
> -d /mnt1/share/opengrok/instances/PROJECT/data \ 
> -W /mnt1/share/opengrok/instances/PROJECT/etc/configuration.xml \ 
> -N /mnt1/share/opengrok/instances/PROJECT/src/ \ 
> -R /mnt1/share/opengrok/common/read_only.xml \ 
> -U http://localhost:8080/PROJECT \ 
> -H -P -S \ 
> --depth 20 \ 
> --progress \ 
> -O on \ 
> --nestingMaximum 5
> 

**Error:**
`Jul 21, 2023 12:39:09 PM org.opengrok.indexer.history.RepositoryFactory getRepository
WARNING: Failed to determineCurrentVersion for /mnt1/share/opengrok/instances/PROJECT/src/kernel/build-2022/prebuilts/ndk-r23: org.eclipse.jgit.errors.MissingObjectException: Missing unknown`

**Version Details:**
> Tomcat version: 10.1.9.0
> OpenGrok version: 1.7.32
> Ubuntu 20.04.6 LTS (Focal Fossa)

I am not sure why the issue is occurring. 
I have the same setup in another server, I could see the historycache is creating over there.

What am I missing here?

#tomcat #opengrok

**Comment by vladak** [comment]
Firstly, the `Missing unknown` message does not seem to be complete, usually it should contain a commit ID. There is not enough context, i.e. I'd like to see other indexer messages related to history index generation, so one might just speculate whether the inability to determine current version of a repository could lead to history not being generated. It seems plausible, though. Is there perhaps complete stack in the logs for the exception ?

Also, I wonder if there are other repositories under the source root and how they fare w.r.t. history cache.

Lastly, what happens if you try to get the HEAD revision for that repository from the command line ? (i.e. `git rev-parse HEAD` or such)

**Comment by mark-nicholson** [comment]
I'm running into a similar issue with the latest version of OpenGrok 1.13.29.  It happens when I run:

`opengrok-reindex-project --printoutput --api_timeout 8 --jar /opengrok/lib/opengrok.jar -U http://localhost:8080/ -P STABLE-5.4 -- --connectTimeout 8 -r dirbased -G -m 256 --leadingWildCards on -c /usr/bin/ctags -U http://localhost:8080/ -H STABLE-5.4`

and shows:

`INFO: Creating history cache for {dir='/opengrok/src/STABLE-5.4',type=git,history=on,historyCache=on,,merge=true,annotationCache=off}
Apr 30, 2025 7:45:59 PM org.opengrok.indexer.history.HistoryGuru lambda$createHistoryCacheReal$36
WARNING: failed to create history cache for {dir='/opengrok/src/STABLE-5.4',type=git,history=on,historyCache=on,,merge=true,annotationCache=off}
org.opengrok.indexer.history.HistoryException: cannot resolve HEAD
	at org.opengrok.indexer.history.GitRepository.accept(GitRepository.java:487)
	at org.opengrok.indexer.history.BoundaryChangesets.getBoundaryChangesetIDs(BoundaryChangesets.java:111)
	at org.opengrok.indexer.history.RepositoryWithHistoryTraversal.doCreateCache(RepositoryWithHistoryTraversal.java:173)
	at org.opengrok.indexer.history.Repository.createCache(Repository.java:403)
	at org.opengrok.indexer.history.HistoryGuru.createHistoryCache(HistoryGuru.java:1078)
	at org.opengrok.indexer.history.HistoryGuru.lambda$createHistoryCacheReal$36(HistoryGuru.java:1122)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
`

The _adjacent_ projects parse just fine.   The ONE difference between those other projects and this one is that the one that fails is a `git worktree`.   I cannot see how that would make a difference though.

In a regular shell, I can run `git rev-parse HEAD` without any problem...  It shows the correct commit-ide.

**Reply by vladak** [reply]
The inability to resolve HEAD with non main Git worktree seems to be a limitation of JGit as explained e.g. here: https://github.com/git-commit-id/git-commit-id-maven-plugin/issues/215

**Reply by vladak** [reply]
JGit [7.0](https://github.com/eclipse-jgit/jgit/wiki/New-and-Noteworthy-7.0) advertises read-only worktree support. OpenGrok currently has 6.6.1.

**Reply by vladak** [reply]
OpenGrok 1.13.32 has JGit 7.2.1, you can try that.

**Reply by mark-nicholson** [reply]
Excellent!  Thanks for the update.  I'll give that a go!

**Reply by mark-nicholson** [reply]
No luck.  I think it is happier, but not quite working.  When opengrok refreshes the configuration.xml, it misses some entries in the worktree case.  I think that prevents it from engaging the indexer.
